Doc/library/asyncio-sync.rst

```diff
@@ -340,6 +341,153 @@ BoundedSemaphore
    .. versionchanged:: 3.10
       Removed the *loop* parameter.
 
+
+Barrier
+=======
+
+   .. versionadded:: 3.11
+
+   A barrier object.  Not thread-safe.
+
+   This class, as a clone of :class:`threading.Barrier`, provides a simple synchronization
+   primitive for use by a fixed number of tasks that need to wait for each other.
+   Each of the tasks tries to pass the barrier by calling the :meth:`~Barrier.wait` method
+   and would block until all of the tasks have made their :meth:`~Barrier.wait` calls.
+   At this point, the tasks are released simultaneously.
+
+   The barrier can be reused any number of times for the same number of tasks.
+
+   .. _asyncio_example_barrier:
+
+   Example::
+
+       import asyncio
+       import os
+       import random
+
+       DELAYS = (1.0, 4.0, 2.0, 0.5)
+       NB = 10
+
+       async def job_done():
+          print(f'{os.linesep}{"All files removed":-^80s}{os.linesep}')
+
+       async def remove_file(barrier, filename):
+          print(f'Remove {filename!r} ...')
+
+          # remove filename (simulation with 'random value' secs sleeping)
+          await asyncio.sleep(random.choice(DELAYS))
+          print(f'\tFile {filename!r} is removed')
+
+          # wait for other files
+          pos = await barrier.wait()
+
+          return pos
+
+       async def remove_folder(barrier, folder):
+          print(f'Waiting for remove all files from folder {folder!r} ...')
+
+          # wait for all removed files
+          async with barrier as pos:
+            # remove folder (simulation with 1 sec sleeping)
+            await asyncio.sleep(1.0)
+            print(f'Folder {folder!r} is now removed ...')
+
+            return pos
+
+       async def main(folder, files, nb_files):
+          # Create a Barrier object with parties equal to nb_files+1
+          # corresponding to the number of all removing items plus the folder,
+          # and an action as a print message.
+          barrier = asyncio.Barrier(nb_files+1, action=job_done)
+
+          # Add remove_file task for each filename.
+          tasks = [asyncio.create_task(remove_file(barrier, f)) for f in files]
+
+          # Add remove_folder task for the folder.
+          tasks.append(asyncio.create_task(remove_folder(barrier, folder)))
+
+          # Wait until folder+filenames are removed.
+          await asyncio.gather(*tasks)
+
+       asyncio.run(main("MyFolder", [f'file{i+1:02d}.txt' for i in range(NB)], NB))
```

Comment:
This example is rather long and contrived. I'd drop it.

Reply:
Ok. Do we need a example ?

Reply:
I proposed an another example